File: src/show_me_tokens/formatters/table.py
"""Human-readable table formatter."""

from __future__ import annotations

from show_me_tokens.models import SessionUsage


def format_table(sessions: list[SessionUsage]) -> str:
    """Render a list of SessionUsage rows as a plain-text table."""
    if not sessions:
        return "No sessions matched the filters."

    headers = [
        "Agent",
        "Session ID",
        "Model",
        "Input",
        "Cache Read",
        "Output",
        "Total",
    ]
    rows: list[list[str]] = []
    for usage in sessions:
        tokens = usage.tokens
        rows.append(
            [
                usage.agent,
                _shorten(usage.session_id, 24),
                usage.model_name or "-",
                _fmt(tokens.input_tokens),
                _fmt(tokens.cache_read_input_tokens),
                _fmt(tokens.output_tokens),
                _fmt(tokens.total_tokens),
            ]
        )

    totals = _compute_totals(sessions)
    rows.append(["-" * len(cell) for cell in rows[0]])
    rows.append(
        [
            "Total",
            "",
            "",
            _fmt(totals["input_tokens"]),
            _fmt(totals["cache_read_input_tokens"]),
            _fmt(totals["output_tokens"]),
            _fmt(totals["total_tokens"]),
        ]
    )

    widths = [len(h) for h in headers]
    for row in rows:
        widths = [max(widths[i], len(row[i])) for i in range(len(headers))]

    lines: list[str] = []
    lines.append("  ".join(headers[i].ljust(widths[i]) for i in range(len(headers))))
    for row in rows:
        lines.append("  ".join(row[i].ljust(widths[i]) for i in range(len(headers))))

    notes = [u.note for u in sessions if u.note]
    if notes:
        lines.append("")
        lines.append("Notes:")
        for note in notes:
            lines.append(f"  - {note}")

    return "\n".join(lines)


def _fmt(value: int | None) -> str:
    if value is None:
        return "-"
    return f"{value:,}"


def _shorten(value: str, max_len: int) -> str:
    if len(value) <= max_len:
        return value
    return value[: max_len - 3] + "..."


def _compute_totals(sessions: list[SessionUsage]) -> dict[str, int | None]:
    input_tokens = sum(u.tokens.input_tokens for u in sessions)
    output_tokens = sum(u.tokens.output_tokens for u in sessions)
    cache_read = sum(
        (u.tokens.cache_read_input_tokens or 0) for u in sessions
    )
    cache_creation = sum(
        (u.tokens.cache_creation_input_tokens or 0) for u in sessions
    )
    total_tokens = sum(
        (u.tokens.total_tokens or 0) for u in sessions
    ) or None
    return {
        "input_tokens": input_tokens,
        "output_tokens": output_tokens,
        "cache_read_input_tokens": cache_read or None,
        "cache_creation_input_tokens": cache_creation or None,
        "total_tokens": total_tokens,
    }

File: src/formatters/table.ts
import path from 'node:path';
import { formatUsd } from '../pricing/huggingface.js';
import { SessionUsage, TokenSummary } from '../shared/types.js';

const HEADERS = [
  'Agent',
  'Session ID',
  'Title',
  'Workspace',
  'Model',
  'Input',
  'Cache Read',
  'Output',
  'Total',
  'HF Cost',
];

export function formatTable(sessions: SessionUsage[]): string {
  if (sessions.length === 0) {
    return 'No sessions matched the filters.';
  }

  const rows: string[][] = [];
  for (const usage of sessions) {
    const tokens = usage.tokens;
    rows.push([
      usage.agent,
      usage.sessionId,
      shorten(usage.title ?? '', 30),
      workspaceName(usage.workspacePath),
      usage.modelName ?? '-',
      fmt(tokens.inputTokens),
      fmt(tokens.cacheReadInputTokens),
      fmt(tokens.outputTokens),
      fmt(tokens.totalTokens),
      formatUsd(usage.pricing?.costUsd),
    ]);
  }

  const totals = computeTotals(sessions);
  rows.push(rows[0].map((cell) => '-'.repeat(cell.length)));
  rows.push([
    'Total',
    '',
    '',
    '',
    '',
    fmt(totals.inputTokens),
    fmt(totals.cacheReadInputTokens),
    fmt(totals.outputTokens),
    fmt(totals.totalTokens),
    formatUsd(computeTotalCostUsd(sessions)),
  ]);

  const widths = HEADERS.map((h) => h.length);
  for (const row of rows) {
    for (let i = 0; i < HEADERS.length; i++) {
      widths[i] = Math.max(widths[i], row[i].length);
    }
  }

  const lines: string[] = [];
  lines.push(
    HEADERS.map((h, i) => h.padEnd(widths[i])).join('  ')
  );
  for (const row of rows) {
    lines.push(row.map((cell, i) => cell.padEnd(widths[i])).join('  '));
  }

  const notes = sessions
    .map((u) => u.note)
    .filter((n): n is string => Boolean(n));
  if (notes.length > 0) {
    lines.push('');
    lines.push('Notes:');
    for (const note of notes) {
      lines.push(`  - ${note}`);
    }
  }

  return lines.join('\n');
}

function fmt(value: number | null): string {
  if (value === null || value === undefined) {
    return '-';
  }
  return value.toLocaleString('en-US');
}

function shorten(value: string, maxLen: number): string {
  if (value.length <= maxLen) {
    return value;
  }
  return value.slice(0, maxLen - 3) + '...';
}

function workspaceName(value: string | null): string {
  if (!value) {
    return 'default';
  }
  const base = path.basename(value);
  return base || 'default';
}

function computeTotalCostUsd(sessions: SessionUsage[]): number | null {
  const total = sessions.reduce(
    (sum, u) => sum + (u.pricing?.costUsd ?? 0),
    0
  );
  return total || null;
}

function computeTotals(sessions: SessionUsage[]): TokenSummary {
  const inputTokens = sessions.reduce(
    (sum, u) => sum + u.tokens.inputTokens,
    0
  );
  const outputTokens = sessions.reduce(
    (sum, u) => sum + u.tokens.outputTokens,
    0
  );
  const cacheRead = sessions.reduce(
    (sum, u) => sum + (u.tokens.cacheReadInputTokens ?? 0),
    0
  );
  const cacheCreation = sessions.reduce(
    (sum, u) => sum + (u.tokens.cacheCreationInputTokens ?? 0),
    0
  );
  const total = sessions.reduce(
    (sum, u) => sum + (u.tokens.totalTokens ?? 0),
    0
  );
  return {
    inputTokens,
    outputTokens,
    cacheReadInputTokens: cacheRead || null,
    cacheCreationInputTokens: cacheCreation || null,
    totalTokens: total || null,
  };
}
